README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

offline_store_upload: Stores api calls locally and uploads when needed. This package can be used in apps where offline functionality is needed.
 

## Usage
 
```dart

import 'package:offline_store_upload/offline_store_upload.dart'; 
  
/// Store API 
MaterialButton(
  child: Text('Store Api'),
  onPressed: () {

    /// initialize a list
    List<OfflineStoreApi> data = [];

    /// add api calls to the list

    data.add(
      OfflineStoreApi(
        api_body: {
          "name": "Flutter",
          "age": 30,
        },
        api_url: url,
        api_type: APITYPE.POST,
      ),
    );

    /// add api calls to the list

    data.add(
      OfflineStoreApi(
        api_body: {
          "name": "Dart",
          "age": 55,
        },
        api_url: url,
        api_type: APITYPE.PUT,
      ),
    );

    /// upload when needed

    OfflineStoreUpload().addData(api_data: data);
  },
),

/// Upload Api
MaterialButton(
  child: Text('UPLOAD API'),
  onPressed: () {

    /// Use this function to upload the locally stored api when needed
    OfflineStoreUpload().upload(token: '');
  },
),

```


CHANGELOG.md
## 0.0.2

* Simple offline storage and api upload for POST and PUT request.

## 0.0.1

* Simple offline storage and api upload.


test/offline_store_upload_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:offline_store_upload/offline_store_upload.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/src/storage.dart
import 'dart:async';
import 'dart:io';
import 'package:path_provider/path_provider.dart';
import 'package:dio/dio.dart';

import '../offline_store_upload.dart';

class OfflineStoreUpload {
  /// to get the local directory path
  Future<String> get _localPath async {
    final directory = await getApplicationDocumentsDirectory();

    return directory.path;
  }

  /// file path
  Future<File> get _localFile async {
    final path = await _localPath;
    return File('$path/offlineStoreUpload.txt');
  }

  /// returns the data that is stored in local file
  Future<String> readData() async {
    try {
      final file = await _localFile;

      // Read the file
      final contents = await file.readAsString();
      print('*** read ****');
      print(contents);
      return contents;
    } catch (e) {
      // If encountering an error, return ''
      return '';
    }
  }

  /// add new api request in queue in local file
  ///
  /// if data alredy exist in file, then
  /// add old data and new data then convert the whole into map and store it into file
  Future<void> addData({
    required List<OfflineStoreApi> api_data,
  }) async {
    String k = await OfflineStoreUpload().readData();

    List<OfflineStoreApi> olddata = [];
    if (k.trim() != '') {
      olddata = offlineStoreApiFromJson(k);
    }

    ///
    List<OfflineStoreApi> newdata = olddata + api_data;
    String newDataToStore = offlineStoreApiToJson(newdata);

    ///
    final file = await _localFile;

    // Write the file
    file.writeAsString('$newDataToStore');
    return;
  }

  /// uploads the api
  ///
  /// if api request returns error then it is store back in the file
  ///
  /// if no error, then that particular api call is removed from the file
  Future<void> upload({required String token}) async {
    String k = await OfflineStoreUpload().readData();
    if (k.trim() == '') return;
    List<OfflineStoreApi> l = offlineStoreApiFromJson(k);

    List<OfflineStoreApi> notUploaded = [];
    for (var i = 0; i < l.length; i++) {
      Response? status = await ApiService().upload(l[i], token);

      if (status == null ||
          (status.statusCode! < 200 || status.statusCode! > 300)) {
        notUploaded.add(l[i]);
      }
    }

    await clear();
    if (notUploaded.isNotEmpty) await addData(api_data: notUploaded);
  }

  /// clear all the data in the file
  Future<void> clear() async {
    final file = await _localFile;

    // Write the file
    file.writeAsString('');
    return;
  }

  /// returns status is there is data present in the local file
  Future<bool> hasDataToUpload() async {
    String k = await OfflineStoreUpload().readData();
    if (k.trim() == '') return false;

    return true;
  }
}


lib/src/network/apiUpload.dart
import 'package:dio/dio.dart';
import 'package:offline_store_upload/src/network/api.dart';

import '../../offline_store_upload.dart';

class ApiService {
  API api = API();
  Future<Response?> upload(OfflineStoreApi data, String token) async {
    Response? response;
    try {
      /// If api call is POST
      if (data.api_type == APITYPE.POST) {
        response = await api.sendRequest.post(
          data.api_url,
          data: data.api_body,
          options: Options(
            headers: {
              'Authorization': 'Bearer $token',
            },
          ),
        );
      }

      /// If api call is PUT
      else if (data.api_type == APITYPE.PUT) {
        response = await api.sendRequest.put(
          data.api_url,
          data: data.api_body,
          options: Options(
            headers: {
              'Authorization': 'Bearer $token',
            },
          ),
        );
      }
      return response;
    } catch (e) {
      print(e);
      return null;
    }
  }
}


lib/src/OfflineStoreUploadModel.dart
import 'dart:convert';

enum APITYPE {
  POST,
  PUT,
}

List<OfflineStoreApi> offlineStoreApiFromJson(String str) =>
    List<OfflineStoreApi>.from(
        json.decode(str).map((x) => OfflineStoreApi.fromJson(x)));

String offlineStoreApiToJson(List<OfflineStoreApi> data) =>
    json.encode(List<dynamic>.from(data.map((x) => x.toJson())));

class OfflineStoreApi {
  OfflineStoreApi({
    required this.api_body,
    // required this.filepath, // TODO: ADD CUSTOM FILE PATH
    required this.api_url,
    required this.api_type,
  });

  Map api_body;
  // String filepath;  // TODO: ADD CUSTOM FILE PATH
  String api_url;
  APITYPE api_type; // POST, PUT

  factory OfflineStoreApi.fromJson(Map<String, dynamic> json) =>
      OfflineStoreApi(
        api_body: json["api_body"],
        // filepath: json["filepath"],
        api_url: json["api_url"],
        api_type: json["api_type"] == 'POST' ? APITYPE.POST : APITYPE.PUT,
      );

  Map<String, dynamic> toJson() => {
        "api_body": api_body,
        // "filepath": filepath,
        "api_url": api_url,
        "api_type": api_type == APITYPE.POST ? 'POST' : 'PUT',
      };
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:offline_store_upload/offline_store_upload.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Offline Store Upload ',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: UploadExample(),
    );
  }
}

class UploadExample extends StatelessWidget {
  String url = 'http://192.168.1.19:8000/api/person/';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            ///
            MaterialButton(
              child: Text('POST 1'),
              onPressed: () {
                List<OfflineStoreApi> data = [];
                data.add(
                  OfflineStoreApi(
                    api_body: {
                      "name": "Hussain",
                      "age": 22,
                    },
                    api_url: url,
                    api_type: APITYPE.POST,
                  ),
                );
                OfflineStoreUpload().addData(api_data: data);
              },
            ),
            SizedBox(height: 20),

            ///
            MaterialButton(
              child: Text('POST 2'),
              onPressed: () {
                List<OfflineStoreApi> data = [];
                data.add(
                  OfflineStoreApi(
                    api_body: {
                      "name": "Flutter",
                      "age": 30,
                    },
                    api_url: url,
                    api_type: APITYPE.POST,
                  ),
                );
                data.add(
                  OfflineStoreApi(
                    api_body: {
                      "name": "Dart",
                      "age": 55,
                    },
                    api_url: url,
                    api_type: APITYPE.POST,
                  ),
                );
                OfflineStoreUpload().addData(api_data: data);
              },
            ),
            SizedBox(height: 20),

            ////
            MaterialButton(
              child: Text('PUT 1'),
              onPressed: () {
                List<OfflineStoreApi> data = [];
                data.add(
                  OfflineStoreApi(
                    api_body: {
                      "name": "Hussain",
                      "age": 20,
                    },
                    api_url: url,
                    api_type: APITYPE.PUT,
                  ),
                );
                OfflineStoreUpload().addData(api_data: data);
              },
            ),
            SizedBox(height: 20),
            MaterialButton(
              child: Text('UPLOAD'),
              onPressed: () {
                OfflineStoreUpload().upload(token: '');
              },
            ),
            SizedBox(height: 20),

            MaterialButton(
              child: Text('DELETE'),
              onPressed: () {
                OfflineStoreUpload().clear();
              },
            ),
          ],
        ),
      ),
    );
  }
}
